lib/dry/system/plugins/zeitwerk.rb

```diff
@@ -0,0 +1,103 @@
+# frozen_string_literal: true
+
+require "dry/system/plugins/zeitwerk/compat_inflector"
+
+module Dry
+  module System
+    module Plugins
+      class Zeitwerk < Module
+        attr_reader :options
+
+        # @api private
+        def initialize(options)
+          @options = options
+          super()
+        end
+
+        # @api private
+        def extended(system)
+          system.before(:configure, &method(:configure_autoloading))
+          system.after(:configure, &method(:setup_autoloader))
+
+          super
+        end
+
+        private
+
+        # Configure autoloading on the container
+        #
+        # @return [self]
+        #
+        # @api private
+        def configure_autoloading(system)
+          require "dry/system/loader/autoloading"
+
+          system.setting :autoloader, reader: true
+          system.config.component_dirs.loader = Dry::System::Loader::Autoloading
+          system.config.component_dirs.add_to_load_path = false
+
+          system
+        end
+
+        # Set a logger
+        #
+        # This is invoked automatically when a container is being configured
+        #
+        # @return [self]
+        #
+        # @api private
+        def setup_autoloader(system)
+          return system if system.registered?(:autoloader)
+
+          if system.config.autoloader
+            system.register(:autoloader, system.config.autoloader)
+          else
+            system.config.autoloader = build_zeitwerk_loader(system)
+          end
+
+          system
+        end
+
+        # Build a zeitwerk loader with the configured component directories
+        #
+        # @return [Zeitwerk::Loader]
+        #
+        # @api private
+        def build_zeitwerk_loader(system)
+          require "zeitwerk"
+
+          loader = options.fetch(:loader) { ::Zeitwerk::Loader.new }
+          loader.tag = system.config.name || system.name
+          loader.logger = method(:puts) if options[:debug]
+          loader.inflector = CompatInflector.new(system.config)
+          push_component_dirs_to_loader(system, loader)
+          loader.setup
+          system.after(:finalize) { loader.eager_load } if eager_load?(system)
+          loader
+        end
+
+        # Add component dirs to the zeitwerk loader
+        #
+        # @return [Zeitwerk::Loader]
+        #
+        # @api private
+        def push_component_dirs_to_loader(system, loader)
+          system.config.component_dirs.each do |dir|
+            raise ZeitwerkAddToLoadPathError, dir if dir.add_to_load_path
```

Comment:
I wonder if we actually want to raise this error at all.

This nothing actually _wrong_ with having a directory in load path and also passing it with Zeitwerk. In that scenario, everything would still work as expected with Zeitwerk.

Given this, and given we already provide the default `add_to_load_path = false` up in `#configure_autoloading`, my feeling is that we should allow users to opt back into this behaviour if they so choose. At that point, they'd have to be doing it for a good reason and we should get out of their way.

What do you think about removing this error?